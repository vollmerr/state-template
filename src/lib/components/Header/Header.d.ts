import * as React from 'react';

import * as types from '../../utils/types';

export interface HeaderProps {
  /** Keeps header fixed to the top */
  fixed?: boolean;

  /** Aligns navigation items */
  align?: 'center'
    | 'left'
    | 'right';

  /** Branding logo */
  brandingLogo?: types.BrandingLogo,

  /** Link to use for 'contact us' link */
  contactLink?: types.ContactLink;

  /** Title to render in place of the home link */
  title: string;

  /** Routes for navigation items */
  routes?: Array<types.Route>;
}

declare class Header extends React.Component<HeaderProps, {}> {}

export default Header;
