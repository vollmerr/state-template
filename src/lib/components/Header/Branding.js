import React from 'react';
import T from 'prop-types';

import A from '../A';

const Branding = ({
  src, alt, to, href,
}) => (
  <div className={'branding'}>
    <div className={'header-organization-banner'}>
      <A to={to} href={href}>
        <img src={src} alt={alt} />
      </A>
    </div>
  </div>
);

Branding.propTypes = {
  /** Source of image */
  src: T.string.isRequired,
  /** Alt description of image */
  alt: T.string.isRequired,
  /** Internal link to navigate to */
  to: T.string,
  /** External link to navigate to */
  href: T.string,
};

Branding.defaultProps = {
  to: '',
  href: '',
};

export default Branding;
