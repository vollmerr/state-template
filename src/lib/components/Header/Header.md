```jsx
const Home = () => <div>in Home</div>;
const Help = () => <div>in Help</div>;

const routes = [
  {
    key: 'home',
    name: 'Home',
    path: '/',
    exact: true,
    icon: 'ca-gov-icon-home',
    component: Home,
  },
  {
    key: 'help',
    name: 'Help',
    path: '/help',
    exact: true,
    icon: 'ca-gov-icon-question-line',
    component: Help,
  },
];

const contactLink = {
  text: 'Contact Link!',
  href: 'https://cdt.ca.gov/support/',
};

<div style={{ overflow: 'auto', display: 'grid' }}>
  <Header 
    fixed={false} 
    align={'right'}
    contactLink={contactLink}
    routes={routes} 
  />
</div>
```
