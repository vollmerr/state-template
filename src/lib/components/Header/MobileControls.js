import React from 'react';
import T from 'prop-types';

import * as propUtils from '../../utils/propTypes';

import Button from '../Button';

const MobileControls = ({ routes, toggleMobileOpen }) => {
  const menuButtonProps = {
    'aria-controls': 'navigation',
    'aria-expanded': 'false',
    baseClass: '',
    className: 'mobile-control toggle-menu pull-right hidden-print',
    iconProps: { name: 'menu', srOnly: 'Menu' },
    onClick: toggleMobileOpen,
    type: 'button',
  };

  return (
    <div className={'mobile-controls'}>
      <span className={'mobile-control-group mobile-header-icons'} />

      {
        Boolean(routes.length) && (
          <div className={'mobile-control-group main-nav-icons pull-right'}>
            <Button {...menuButtonProps} />
          </div>
        )
      }
    </div>
  );
};

MobileControls.propTypes = {
  routes: T.arrayOf(propUtils.route),
  toggleMobileOpen: T.func.isRequired,
};

MobileControls.defaultProps = {
  routes: [],
};

export default MobileControls;
