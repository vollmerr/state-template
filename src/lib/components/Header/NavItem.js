import React from 'react';
import { withRouter } from 'react-router-dom';

import A from '../A';

/**
 * A single navigation item
 * @param {string} to       - path to navigate to
 * @param {array} children  - children to render
 * @param {func} onClick    - fired when item is clicked on
 * @param {string} hash     - current hash name from redux
 */
const NavItem = ({
  name,
  path,
  href,
  icon,
  toggleMobileOpen,
  hash,
  onClick,
  location,
  ...props
}) => {
  const isActive = location.pathname === path;
  const activeClass = isActive ? 'active' : '';
  const mappedOnClick = (event) => {
    if (onClick) onClick(event);
    toggleMobileOpen(event);
  };

  return (
    <li className={`nav-item ${activeClass}`}>
      <A
        to={path}
        href={href}
        className={'first-level-link'}
        onClick={mappedOnClick}
      >
        <span className={icon} aria-hidden={'true'} />
        {name}
      </A>
    </li>
  );
};

export default withRouter(NavItem);
