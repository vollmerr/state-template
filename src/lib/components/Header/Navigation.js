import React from 'react';
import T from 'prop-types';

import * as propUtils from '../../utils/propTypes';
import NavItem from './NavItem';

const Navigation = ({
  isMobileOpen,
  routes,
  toggleMobileOpen,
}) => {
  const closedClass = isMobileOpen ? '' : 'mobile-closed';

  if (!routes.length) {
    return null;
  }

  return (
    <nav id={'navigation'} className={`main-navigation singlelevel auto-highlight ${closedClass}`}>
      <ul id={'nav_list'} className={'top-level-nav'}>
        {
          routes.map((route) => (
            <NavItem {...route} toggleMobileOpen={toggleMobileOpen} />
          ))
        }
      </ul>
    </nav>
  );
};

Navigation.propTypes = {
  isMobileOpen: T.bool.isRequired,
  routes: T.arrayOf(propUtils.route).isRequired,
  toggleMobileOpen: T.func.isRequired,
};

export default Navigation;
