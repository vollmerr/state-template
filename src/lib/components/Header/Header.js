import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';
import debounce from 'lodash.debounce';

import * as propUtils from '../../utils/propTypes';
import odiLogo from '../../images/office-of-digital-innovation-logo.png';
import Overlay from '../Overlay';

import Branding from './Branding';
import MobileControls from './MobileControls';
import Navigation from './Navigation';
import Search from './Search';
import SettingsBar from './SettingsBar';
import UtilityHeader from './UtilityHeader';

/**
 * Renders a header that handles responsive viewports,
 * compacting on scroll, opening settings, and
 * generating navigation.
 */
class Header extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      isCompact: false,
      isMobileOpen: false,
      isSettingsOpen: false,
    };
    this.scrollEvent = debounce(this.onScroll, 200);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.scrollEvent);
    this.setSettings();
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.scrollEvent);
  }

  // set setttings user has set (in SettingsBar)
  setSettings = () => {
    document.body.style.zoom = Number(localStorage.getItem('zoom'));
    if (localStorage.getItem('highContrast')) {
      document.getElementsByTagName('html')[0].classList.add('high-contrast');
    }
  }

  // handles compacting header when scrolled down
  onScroll = () => {
    const { isCompact } = this.state;
    const newIsCompact = window.scrollY > 130;

    if (isCompact !== newIsCompact) {
      this.setState({ isCompact: newIsCompact });
    }
  }

  // handles toggling if the mobile menu is open
  toggleMobileOpen = () => {
    const { isMobileOpen } = this.state;
    this.setState({ isMobileOpen: !isMobileOpen });
  }

  // handles toggling if the settings menu is open
  toggleSettingsOpen = () => {
    const { isSettingsOpen } = this.state;
    this.setState({ isSettingsOpen: !isSettingsOpen });
  }

  render() {
    const {
      fixed,
      align,
      brandingLogo,
      routes,
      contactLink,
    } = this.props;

    const {
      isCompact,
      isMobileOpen,
      isSettingsOpen,
    } = this.state;

    const filteredRoutes = routes.filter(x => !x.hidden);
    const cn = classNames([
      'global-header',
      { fixed },
      { compact: isCompact },
      align,
    ]);

    return (
      <>
        <header role={'banner'} className={cn}>
          <UtilityHeader
            toggleSettingsOpen={this.toggleSettingsOpen}
            contactLink={contactLink}
          />

          <SettingsBar
            toggleSettingsOpen={this.toggleSettingsOpen}
            isSettingsOpen={isSettingsOpen}
          />

          <Branding {...brandingLogo} />

          <MobileControls
            toggleMobileOpen={this.toggleMobileOpen}
            routes={filteredRoutes}
          />

          <div className={'navigation-search'}>
            <Navigation
              isMobileOpen={isMobileOpen}
              toggleMobileOpen={this.toggleMobileOpen}
              routes={filteredRoutes}
            />
            <div id={'head-search'} className={'search-container'}>
              <Search />
            </div>
          </div>


          <div className={'header-decoration'} />
        </header>

        <Overlay
          isMobile
          show={isMobileOpen}
          onClick={this.toggleMobileOpen}
        />
      </>
    );
  }
}

Header.propTypes = {
  /** Keeps header fixed to the top */
  fixed: T.bool,
  /** Aligns navigation items */
  align: T.oneOf(['center', 'left', 'right']),
  /** Branding logo */
  brandingLogo: propUtils.brandingLogo,
  /** Contact us link */
  contactLink: propUtils.contactLink,
  /** Routes for navigation items */
  routes: T.arrayOf(propUtils.route),
};

Header.defaultProps = {
  fixed: true,
  align: 'center',
  brandingLogo: {
    src: odiLogo,
    alt: 'Office of Digital Innovation logo',
    href: 'https://cdt.ca.gov/',
  },
  contactLink: {
    text: 'Contact Us',
    href: 'https://cdt.ca.gov/support/',
  },
  routes: [],
};

export default Header;
