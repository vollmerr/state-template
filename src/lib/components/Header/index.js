import Branding from './Branding';
import Header from './Header';
import LocationBar from './LocationBar';
import MobileControls from './MobileControls';
import Navigation from './Navigation';
import NavItem from './NavItem';
import Search from './Search';
import SettingsBar from './SettingsBar';
import UtilityHeader from './UtilityHeader';

export {
  Branding,
  LocationBar,
  MobileControls,
  Navigation,
  NavItem,
  Search,
  SettingsBar,
  UtilityHeader,
};

export default Header;
