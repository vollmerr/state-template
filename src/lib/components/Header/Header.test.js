import React from 'react';
import { shallow } from 'enzyme';

import { Header as fromLib } from '../..';
import Header from './Header';

let wrapper;
describe('Header', () => {
  beforeEach(() => {
    wrapper = shallow(<Header />);
  });

  it('should be exported as part of the library', () => {
    expect(fromLib).toEqual(Header);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
