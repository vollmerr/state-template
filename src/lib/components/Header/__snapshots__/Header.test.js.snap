// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<Fragment>
  <header
    className="global-header fixed center"
    data-test="header"
    role="banner"
  >
    <UtilityHeader
      contactLink={
        Object {
          "href": "https://cdt.ca.gov/support/",
          "text": "Contact Us",
        }
      }
      toggleSettingsOpen={[Function]}
    />
    <SettingsBar
      isSettingsOpen={false}
      toggleSettingsOpen={[Function]}
    />
    <Branding
      alt="Office of Digital Innovation logo"
      href="https://cdt.ca.gov/"
      src="office-of-digital-innovation-logo.png"
      to=""
    />
    <MobileControls
      routes={Array []}
      toggleMobileOpen={[Function]}
    />
    <div
      className="navigation-search"
    >
      <Navigation
        isMobileOpen={false}
        routes={Array []}
        toggleMobileOpen={[Function]}
      />
      <div
        className="search-container"
        id="head-search"
      >
        <Search />
      </div>
    </div>
    <div
      className="header-decoration"
    />
  </header>
  <Overlay
    className="header-overlay"
    isMobile={true}
    onClick={[Function]}
    show={false}
  />
</Fragment>
`;
